### PR TITLE
Expand macros in trait and impl bodies

### PR DESCRIPTION
**Pull request by vlad20012**
Now we can expand macros in this context:
```rust
impl Foo {
    foo!();
    bar!();
}
```
and can resolve method generated by these macros.
Also fixed some inspections to work properly with members defined by a macro.

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt [hunk 1]
```diff
@@ -80,28 +79,26 @@ private fun insertNewTraitMembers(selected: Collection<RsAbstractable>, existing
         }
 
         val anchor = indexedExistingMembers
-            .map { IndexedValue(it.index, it.value.first) }
-            .lastOrNull()
+            .map {
+                val member = it.value.first
+                IndexedValue(it.index, member.expandedFromRecursively ?: member)
+            }.lastOrNull()
             ?: IndexedValue(-1, existingMembers.lbrace) ?: return
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt]: This is strange `return`; do we really need it? Constructor cannot return nullable value (I guess)

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsMembers.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+/** Returns all members including those produced by macros */
+val RsMembers.expandedMembers: List<RsAbstractable>
+    get() {
+        val members = mutableListOf<RsAbstractable>()
+
+        for (member in childrenSequence) {
+            when (member) {
+                is RsAbstractable -> members.add(member)
+                is RsMacroCall -> member.processExpansionRecursively {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsMembers.kt]: Maybe we could add addditional function to this place, so this `false` in the air will not look so strange?
Like this: 
```kotlin
private fun RsMacroCall.collectAbstractableMembersRecursively(): List<RsAbstractable> {
    val members = mutableListOf<RsAbstractable>()

    this.processExpansionRecursively { 
        if (it is RsAbstractable) {
            members.add(it)
        }

        false
    }

    return members
}
```
and then use it like 
```kotlin
... -> members.addAll(member.collectAbstractableMembersRecursively())
```

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsMacroCall.kt [hunk 3]
```diff
@@ -53,6 +54,21 @@ val RsMacroCall.expansion: List<ExpansionResult>?
         expandMacro(this)
     }
 
+fun RsMacroCall.processExpansionRecursively(processor: (ExpansionResult) -> Boolean): Boolean =
+    processExpansionRecursively(processor, 0)
+
+private fun RsMacroCall.processExpansionRecursively(processor: (ExpansionResult) -> Boolean, recursionDepth: Int): Boolean {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsMacroCall.kt]: Same here: we can improve this function by using `any` method of the list, like this:
```kotlin
private fun RsMacroCall.processExpansionRecursively(processor: (ExpansionResult) -> Boolean, recursionDepth: Int): Boolean {
    if (recursionDepth > DEFAULT_RECURSION_LIMIT) return true

    return expansion.orEmpty().any { it.processRecursively(processor, recursionDepth) }
}

private fun ExpansionResult.processRecursively(processor: (ExpansionResult) -> Boolean, recursionDepth: Int) = when (this) {
    is RsMacroCall -> this.processExpansionRecursively(processor, recursionDepth + 1)
    else -> processor(this)
}
```

This way there would be only one return point and it is easier to read (imo)

File: src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt [hunk 4]
```diff
@@ -80,28 +79,26 @@ private fun insertNewTraitMembers(selected: Collection<RsAbstractable>, existing
...
         existingMembersWithPosInTrait.add(anchor.index + 1, Pair(addedMember, posInTrait))
 
         // If the newly added item is a function, we add an extra line between it and each of its siblings.
-        val previousAbstractable = addedMember.leftSiblings.find { it is RsAbstractable }
-        if (previousAbstractable != null) {
-            if (previousAbstractable is RsFunction || addedMember is RsFunction) {
-                val whitespaces = createExtraWhitespacesAroundFunction(previousAbstractable, addedMember)
-                existingMembers.addBefore(whitespaces, addedMember)
-            }
+        val prev = addedMember.leftSiblings.find { it is RsAbstractable || it is RsMacroCall }
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt]: Those two pieces are almost identical, maybe we should extract a function here

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsItemsOwner.kt [hunk 5]
```diff
@@ -45,7 +44,7 @@ fun RsItemsOwner.processExpandedItems(processor: (RsItemElement) -> Boolean): Bo
     for (psi in itemsAndMacros) {
```
Comment (fedochet) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsItemsOwner.kt]: Forgot to add that here we can also use `any` method:
```kotlin
fun RsItemsOwner.processExpandedItems(processor: (RsItemElement) -> Boolean): Boolean {
    return itemsAndMacros.any { it.processItem(processor) }
}

private fun RsElement.processItem(processor: (RsItemElement) -> Boolean) = when (this) {
    is RsMacroCall -> this.processExpansionRecursively { it is RsItemElement && processor(it) }

    is RsItemElement -> processor(this)

    else -> false
}

```

File: src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt (hunk 4 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt]: There are a lot of differences: leftSiblings/rightSiblings prev/nex/addedMember order in `createExtraWhitespacesAroundFunction` call parameters, addBefore/addAfter. So I don't know how to generalize it.

Comment (fedochet) [reply src/main/kotlin/org/rust/lang/refactoring/implementMembers/impl.kt]: You're right, missed that

**Comment by vlad20012** [comment]
bors r=fedochet

**Pull request merged**